operator/api/vectorized/v1alpha1/cluster_types.go

```diff
@@ -1418,3 +1419,155 @@ func (r *Cluster) GetDecommissionBrokerID() *int32 {
 func (r *Cluster) SetDecommissionBrokerID(id *int32) {
 	r.Status.DecommissioningNode = id
 }
+
+// getNodePoolsFromSpec returns the NodePools defined in the spec.
+// This contains the primary NodePool (driven by spec.replicas for example), and the
+// NodePools driven by the nodePools field.
+func (r *Cluster) getNodePoolsFromSpec() []NodePoolSpec {
+	out := make([]NodePoolSpec, 0)
+	if r.Spec.Replicas != nil {
+		defaultNodePool := NodePoolSpec{
+			Name:         nodepools.DefaultNodePoolName,
+			Replicas:     r.Spec.Replicas,
+			Tolerations:  r.Spec.Tolerations,
+			NodeSelector: r.Spec.NodeSelector,
+			Storage:      r.Spec.Storage,
+			Resources:    r.Spec.Resources,
+		}
+		if r.Spec.CloudStorage.CacheStorage != nil {
+			defaultNodePool.CloudCacheStorage = *r.Spec.CloudStorage.CacheStorage
+		}
+		out = append(out, defaultNodePool)
+	}
+	out = append(out, r.Spec.NodePools...)
+
+	return out
+}
+
+func (r *Cluster) CalculateCurrentReplicas() int32 {
+	var result int32
+	for _, np := range r.Status.NodePools {
+		result += np.CurrentReplicas
+	}
+	return result
+}
+
+func (r *Cluster) GetNodePools(ctx context.Context, k8sClient client.Reader) ([]*NodePoolSpecWithDeleted, error) {
+	var nodePoolsWithDeleted []*NodePoolSpecWithDeleted
+
+	nps := r.getNodePoolsFromSpec()
+	for i := range nps {
+		np := nps[i]
+
+		nodePoolsWithDeleted = append(nodePoolsWithDeleted, &NodePoolSpecWithDeleted{
+			NodePoolSpec: np,
+		})
+	}
+
+	// Also add "virtual NodePools" based on StatefulSets found.
+	// These represent deleted NodePools - they will not show up in spec.NodePools.
+	var stsList appsv1.StatefulSetList
+	err := k8sClient.List(context.TODO(), &stsList)
+	if err != nil {
+		return nil, err
+	}
+outer:
+	for i := range stsList.Items {
+		sts := stsList.Items[i]
+
+		if !strings.HasPrefix(sts.Name, r.Name) {
+			continue
+		}
+
+		for _, ownerRef := range sts.OwnerReferences {
+			if ownerRef.UID != r.UID {
+				continue outer
+			}
+		}
+
+		var npName string
+		if strings.EqualFold(r.Name, sts.Name) {
+			npName = nodepools.DefaultNodePoolName
+		} else {
+			npName = sts.Name[len(r.Name)+1:]
+		}
+
+		// Have seen it in NodePoolSpec
+		if slices.ContainsFunc(nps, func(np NodePoolSpec) bool {
+			return np.Name == npName
+		}) {
+			continue
+		}
+
+		replicas := sts.Spec.Replicas
+		if st, ok := r.Status.NodePools[npName]; ok {
+			replicas = &st.CurrentReplicas
+		}
```

Comment:
this is a good question. let me check if we can remove it, i am not 100% sure anymore why it is necessary, but i think it is for a reason. i think it's because the normal way of translating currentReplicas to replicas does not work with these "virtual" pools, but i will double check.